lib/sudoku.rb

```diff
@@ -0,0 +1,133 @@
+class Board
+    attr_reader :array
+
+    def initialize(array)
+        @array = array.map do |row|
+            row.map { |element| CellValue.new(element, !element.zero?)}
+        end 
+    end
+
+    def ==(item)
+        array.each_with_index do |row, x| 
+            row.each_with_index do |element, y|
+                if element != item.array[x][y]
+                    return false
+                end
+            end
+        end
+        true
+    end
+
+    def columns
+        column_array = array.transpose
+    end
+    
+    def filled?
+        array.none? { |a| a.include?(0) }
+    end
+
+    def valid?
+        Validator.new(self).valid?
+    end
+
+    def subgroups
+        array.each_slice(3).map{|stripe| stripe.transpose.each_slice(3).map{|chunk| chunk.transpose}}.flatten(1)
+    end
+
+    def to_s
+        array.map! { |word| word.join(' ') }
+        array.map! { |word| word.insert(6, "|") }
+        array.map! { |word| word.insert(13, "|") }
+        array.map! { |word| word.insert(-1, "\n") }
+        array1 = array.insert(3, "------+------+------\n" )
+        array2 = array.insert(7, "------+------+------\n" )
+    end
+
+    def insert(x, y, val)
+        if array[x][y].locked
+            raise PermissionDeniedError.new("Šo vērtību nevar izmainīt")
+        elsif (1..9).include?(val)
+            masiivs = to_a
+            masiivs[x][y] = val
+            masiivs
+        else
+            raise PermissionDeniedError.new("Kļūda! Ievadīt var tikai 1-9")
+        end
+    end
+
+    def insert!(x, y, val)
+        if array[x][y].locked
+            raise PermissionDeniedError.new("Šo vērtību nevar izmainīt")
+        elsif (1..9).include?(val)
+            array[x][y].value = val
+            array
+        else
+            raise PermissionDeniedError.new("Kļūda! Ievadīt var tikai 1-9")
+        end
+    end
```

Comment:
Šīm metodēm baigi daudz lietas atkārtojās, vienīgais kas atšķiras ir darbības iekš `elsif`. 

Parasti mēģinam neatkārtot kodu, jo pēc tam, kad kaut kas jāmaina, ir grūtāk to izdarīt, jo kods ir vairākās vietās.

* Palasi pirmo daļu par blokiem(vari arī pārējās, bet bloki ir tie, kas šeit ir aktuāli) https://blog.appsignal.com/2018/09/04/ruby-magic-closures-in-ruby-blocks-procs-and-lambdas.html 
* Izveido jaunu metodi, kas saņem bloku un veic šo pašu `if-elsif-else` pārbaudi. Aizvieto `elsif` daļu ar bloku
* Nomaini šīs metodes, lai tās izsauc jauno un kā bloku padod katra savu `elsif` loģiku